### PR TITLE
Support VB6 IDE

### PR DESCRIPTION
**Pull request by mansellan**
Further work on #298:

- Abstracted all VBA dependencies
- Removed all Interop Assembly references except for Rubberduck.VBEditor
- Refactored static events for VBProject and VBComponent into abstract base classes
- Renamed Office.dll to Office12.dll, added Office8.dll

### CODE REVIEW

File: RetailCoder.VBE/API/VBA/ParserState.cs [hunk 1]
```diff
@@ -10,14 +10,15 @@
...
+using Rubberduck.VBEditor.SafeComWrappers.VB;
+using Rubberduck.VBEditor.SafeComWrappers.VB.Abstract;
 
 namespace Rubberduck.API.VBA
 {
     [ComVisible(true)]
     public interface IParserState
     {
-        void Initialize(Microsoft.Vbe.Interop.VBE vbe);
+        void Initialize(object vbe);
```
Comment (retailcoder) [inline RetailCoder.VBE/API/VBA/ParserState.cs]: Why not `SafeComWrappers.VB.Abstract.IVBE`?

File: RetailCoder.VBE/API/VBA/ParserState.cs [hunk 2]
```diff
@@ -48,21 +49,21 @@ public sealed class ParserState : IParserState, IDisposable
...
-        private VBE _vbe;
+        private IVBE _vbe;
 
         public ParserState()
         {
             UiDispatcher.Initialize();
         }
 
-        public void Initialize(Microsoft.Vbe.Interop.VBE vbe)
+        public void Initialize(object vbe)
```
Comment (retailcoder) [inline RetailCoder.VBE/API/VBA/ParserState.cs]: ditto

File: Rubberduck.SourceControl/SourceControlProviderBase.cs [hunk 3]
```diff
@@ -120,7 +119,7 @@ public virtual void Undo(string filePath)
         {
             if (File.Exists(filePath))
             {
-                var component = Project.VBComponents.OfType<VBComponent>().FirstOrDefault(f => f.Name == filePath.Split('.')[0]);
```
Comment (retailcoder) [inline Rubberduck.SourceControl/SourceControlProviderBase.cs]: Wah, blast from the past! You've found a relic that predates the wrappers! 😮

File: Rubberduck.VBEEditor/SafeComWrappers/VB/VBComponentFactory.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Rubberduck.VBEditor.SafeComWrappers.VB
+{
+    public static class VBComponentFactory
+    {
+        public static IVBComponent Create(object vbComponent)
+        {
+            switch (vbComponent)
+            {
+                case VBAIA.VBComponent _:
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/SafeComWrappers/VB/VBComponentFactory.cs]: what's that underscore for?

File: Rubberduck.VBEEditor/SafeComWrappers/VB/VBProjectFactory.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    public static class VBProjectFactory
+    {
+        public static IVBProject Create(object project)
+        {
+            switch (project)
+            {
+                case VBAIA.VBProject vbap:
+                    return new VBA.VBProject(vbap);
+
+                case VB6IA.VBProject vb6p:
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/SafeComWrappers/VB/VBProjectFactory.cs]: nevermind. that's C#7?

File: Rubberduck.VBEEditor/SafeComWrappers/VB/VbeFactory.cs [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    public static class VBEFactory
+    {
+        public static IVBE Create(object application)
+        {
+            switch (application)
+            {
+                case VBAIA.VBE vbae:
+                    return new VBA.VBE(vbae);   
+                    
+                case VB6IA.VBE vb6e:
```
Comment (retailcoder) [inline Rubberduck.VBEEditor/SafeComWrappers/VB/VbeFactory.cs]: I think I'd prefer seeing `case VB6IA.VBE ide:` in both branches. Also I'll have to hold myself back very hard to not rename `VB6IA` to just `VB6` and `VBAIA` to `VBA`.

File: RetailCoder.VBE/API/VBA/ParserState.cs (hunk 1 above)
Comment (mansellan) [reply RetailCoder.VBE/API/VBA/ParserState.cs]: The function takes a raw COM object, then wraps it into an IVBE - previously this was a VBA VBE, now uses a factory method to remove the dependency. There's no interface shared between VBA and VB6 COM interop types, hence object.

File: RetailCoder.VBE/API/VBA/ParserState.cs (hunk 2 above)
Comment (mansellan) [reply RetailCoder.VBE/API/VBA/ParserState.cs]: ditto

File: Rubberduck.VBEEditor/SafeComWrappers/VB/VBComponentFactory.cs (hunk 4 above)
Comment (mansellan) [reply Rubberduck.VBEEditor/SafeComWrappers/VB/VBComponentFactory.cs]: switching by type is a C#7 feature, but it requires a variable assignment. The underscore is 'meh, don't care about the assign'

File: RetailCoder.VBE/API/VBA/ParserState.cs (hunk 1 above)
Comment (Vogel612) [reply RetailCoder.VBE/API/VBA/ParserState.cs]: @mansellan seems like a good place to add a "why" comment to help future readers, then

File: Rubberduck.VBEEditor/SafeComWrappers/VB/VbeFactory.cs (hunk 6 above)
Comment (mansellan) [reply Rubberduck.VBEEditor/SafeComWrappers/VB/VbeFactory.cs]: changed to ide. WRT to the IA - at one point during construction I was working with both COM and Wrapper objects, and wanted the suffix to help disambiguate. But now the namespaces are sorted, it seems that's no longer an issue, so I'll remove the IA litter.

File: RetailCoder.VBE/API/VBA/ParserState.cs (hunk 1 above)
Comment (mansellan) [reply RetailCoder.VBE/API/VBA/ParserState.cs]: will do :-)

File: Rubberduck.VBEEditor/SafeComWrappers/VB/VBProjectFactory.cs (hunk 5 above)
Comment (Vogel612) [reply Rubberduck.VBEEditor/SafeComWrappers/VB/VBProjectFactory.cs]: Yes. Pattern matching or Tuple deconstruction. Depends on how exactly it's implemented...

File: Rubberduck.VBEEditor/SafeComWrappers/VB/VbeFactory.cs (hunk 6 above)
Comment (retailcoder) [reply Rubberduck.VBEEditor/SafeComWrappers/VB/VbeFactory.cs]: Would `VBA.Interop` and `VB6.Interop` work?

File: RetailCoder.VBE/API/VBA/ParserState.cs (hunk 1 above)
Comment (retailcoder) [reply RetailCoder.VBE/API/VBA/ParserState.cs]: Gah, just noticed the namespace - of course it takes an interop VBE object; that method is meant to be invoked by VBA/VB6 code! Good call with the factory.

**Pull request merged**